### ISSUE TITLE
parameterized derived atom without caching or deepEqual

### ISSUE THREAD

**Issue by ahfarmer** [pmndrs/jotai#690]
## Overview

Looking for a way to have a "parameterized derived atom", similar to an atomFamily but without caching. This is part of my overall strategy of moving ALL my state into jotai ;)

## Example Setup

As an example, imagine you have a list of 'events' that are fetched from a server. Each event has a timestamp and some other arbitrary data.

```javascript
const allEventsAtom = atom(null)
allEventsAtom.onMount = (setAtom) => {
  fetch("/eventspath").then(response => {
    setAtom(response.events);
  });
}
```

## Example Goal

I want to be able to access a subset of these events. Currently I have two approaches, neither of which I like: hooks and atomFamilies.

## Example Solution 1: Hooks

```javascript
function useEvents(startTime, endTime) {
  const allEvents = useAtomValue(allEventsAtom);
  return allEvents.filter(x => x.timestamp > startTime && x.timestamp < endTime);
}
```

*Problem with hooks:* Not usable in other atoms.

## Example Solution 2: Atom Family

```javascript
const eventsFamily = atomFamily(([start, end]) => atom(get => {
  const allEvents = get(allEventsAtom);
  return allEvents.filter(x => x.timestamp > startTime && x.timestamp < endTime);
}, deepEqual)
```

*Problem with atom families:* Memory leak! Every time start or end changes, the atom family will add a new atom to its `Map`. Start/end may change with high frequency, for instance while the user is dragging a slider.

## Other Ideas

1. Use `myFamily.setShouldRemove(() => true)`. Will this work? Does `shouldRemove` get called before or after an atom is fetched from the cache? Perhaps this would cause infinite recalculations.

2. Write a different version of `atomFamily` that only caches a single value instead of a whole Map of values. I believe this works, but that single cached value sticks around forever, and I still have to do `deepEqual`.

In short, I would ideally want to have something just like a hook: no `deepEqual` and no caching, but still usable in other `atom`s. Is this feasible?

**Comment by dai-shi** [comment]
Could you elaborate how you would like to reuse the atom? Because otherwise my suggestion is this:

```js
const createEventsAtom = (startTime, endTime) => atom(get => {
  const allEvents = get(allEventsAtom)
  return allEvents.filter(x => x.timestamp > startTime && x.timestamp < endTime)
})

function useEvents(startTime, endTime) {
  const filteredEvents = useAtomValue(
    useMemo(() => createEventsAtom(startTime, endTime), [startTime, endTime])
  );
  return filteredEvents
}
```

---

As for `setShouldRemove`, `shouldRemove` is called for all existing atoms when `setShouldRemove` is called, and for a specific atom before it returns it. So, technically, you can call `setShouldRemove` periodically.

**Comment by ahfarmer** [comment]
Okay, let me explore that approach. I'll use your `createEventsAtom` but not your `useEvents` hook because I want to an atom derived from the `allEvents` atom. I want a `relevantEventsAtom` which can alternately return one of two lists of events, based on a boolean atom:


```javascript
const createEventsAtom = (startTime, endTime) => atom(get => {
  const allEvents = get(allEventsAtom)
  return allEvents.filter(x => x.timestamp > startTime && x.timestamp < endTime)
})

const liveAtom = atom(false)

const liveEventAtom = atom({ timestamp: Date.now() })

const relevantEventsAtom = atom(get => {
  const live = get(liveAtom);
  if (live) {
    return [get(liveEventAtom)];
  }
  const [start, end] = get(eventRangeAtom)
  return get(createEventsAtom(start, end))
})
```

This works, except that garbage collecting all those atoms I'm creating will be a pain right?

**Comment by dai-shi** [comment]
Hmm, that usage doesn't help for anything.

```js
const relevantEventsAtom = atom(get => {
  const live = get(liveAtom);
  if (live) {
    return [get(liveEventAtom)];
  }
  const [start, end] = get(eventRangeAtom)
  const allEvents = get(allEventsAtom)
  return allEvents.filter(x => x.timestamp > start && x.timestamp < end)
})
```

So, no need to create an intermediate atom.

> Start/end may change with high frequency, for instance while the user is dragging a slider.

I wonder you would even want caching in this case?

Feels like I misunderstand the problem space.

**Comment by ahfarmer** [comment]
I'm not explaining myself clearly - and I think there is likely nothing for jotai to improve here. I'll re-open or do a new issue if I see a problem and can arrange my thoughts more clearly.

For now I think you are right - I don't need the intermediate atom. If necessary I can just write a static function that would get used in two places: a hook, and an atom. Thanks again!
